### PR TITLE
Add support for building Python 3 wheels and the starforge-recipes repo

### PR DESCRIPTION
**Pull request by natefoo**
New wheels and changes to build new versions of existing wheels should go in [galaxyproject/starforge-recipes](https://github.com/galaxyproject/starforge-recipes/).

### CODE REVIEW

File: starforge/commands/cmd_wheel.py [hunk 1]
```diff
@@ -107,39 +137,38 @@ def cli(ctx, wheels_config, osk, sdist, docker, qemu, wheel, qemu_port,
...
-            if exists(name):
-                break
-        else:
-            msg = "Possible sdists missing, build failed?"
-            if exit_on_failure:
-                fatal(msg)
-            else:
-                warn(msg)
+        wheel_config.set_purepy(False)
+        # sets universal
```
Comment (nsoranzo) [inline starforge/commands/cmd_wheel.py]: `# sets universal to False`

File: starforge/commands/cmd_wheel.py [hunk 2]
```diff
@@ -107,39 +137,38 @@ def cli(ctx, wheels_config, osk, sdist, docker, qemu, wheel, qemu_port,
...
+        debug("Detected wheel type: %s", wheel_type)
+    if wheel_type is None:
+        fatal("ERROR: Unable to determine wheel type of '%s', set `purepy`, `universal`, and/or `imageset` in wheel "
+              "config", wheel_config.name)
+    imageset = TYPE_IMAGESET_MAP[wheel_type]
+    debug("Using imageset: %s", wheel_config.configured_imageset or imageset)
+    wheel_config.set_imageset(imageset=imageset)
+    if wheel_type == UNIVERSAL:
+        wheel_config.set_universal(True)
+        # sets purepy
```
Comment (nsoranzo) [inline starforge/commands/cmd_wheel.py]: `# sets purepy to True`

File: starforge/forge/wheels.py [hunk 3]
```diff
@@ -214,10 +214,14 @@ def bdist_wheel(self, output=None, uid=-1, gid=-1):
 
         chdir(join(build, root))
 
-        if self.wheel_config.insert_setuptools or self.image.plat_specific or not check_setup():
+        insert_setuptools = self.wheel_config.insert_setuptools
+        if not insert_setuptools and insert_setuptools is not False:
```
Comment (nsoranzo) [inline starforge/forge/wheels.py]: Is this equivalent to the simpler `if insert_setuptools is None:` ?

File: starforge/packaging/setup.py [hunk 4]
```diff
@@ -90,4 +90,7 @@ def wheel_type(package_dir=None):
 
 def _check_output(cmd, cwd=None):
     debug('Executing in %s: %s', cwd, ' '.join(cmd))
-    return check_output(cmd, cwd=cwd)
+    out = check_output(cmd, cwd=cwd)
+    if sys.version_info >= (3,):
```
Comment (nsoranzo) [inline starforge/packaging/setup.py]: I think it should be fine to always `decode()`, also on Python 2.7.

File: starforge/forge/wheels.py (hunk 3 above)
Comment (natefoo) [reply starforge/forge/wheels.py]: Yeah, not sure what gymnastics I engaged in to get this.

File: starforge/commands/cmd_test_wheel.py [hunk 5]
```diff
@@ -38,8 +38,10 @@ def cli(ctx, wheels_config, osk, image, qemu_port, wheel):
         fatal("At least one image must be specified")
     for i in image:
         try:
+            info("Testing wheel on image '%s': %s", i, wheel)
             for forge in build_forges(ctx.config, wheels_config, wheel, image=i, osk_file=osk, qemu_port=qemu_port):
                 names = forge.get_expected_names()
+                debug("Expecting wheel files:\n  %s", '\n  '.join(names))
```
Comment (nsoranzo) [inline starforge/commands/cmd_test_wheel.py]: This is never reached for purepy packages like fcsparser if the imageset and the wheel type are not specified in `meta.yaml`, need to try to autodetect the type as in cmd_wheel.py ?

**Comment by nsoranzo** [comment]
@natefoo Time to merge this?

**Pull request merged**